File: Plus One/plus-one.php
<?php
/**
 * Created by PhpStorm.
 * User: kovalevcon
 * Date: 02.05.19
 * Time: 19:05
 */

/**
 * Plus one function
 *
 * @param array[int] $digits
 * @return array
 */
function plusOne(array $digits): array
{
    for ($i = count($digits) - 1; $i >= 0; $i--) {
        if ($digits[$i] < 9) {
            $digits[$i]++;
            return $digits;
        }
        $digits[$i] = 0;
    }

    if ($digits[0] === 0) {
        array_unshift($digits, 1);
    }

    return $digits;
}

echo "<pre>plusOne([1, 2, 3])\r\n";
var_dump(plusOne([1, 2, 3])) ; // [1, 2, 4]
echo "</pre>";

echo "<pre>plusOne([4, 3, 2, 1])\r\n";
var_dump(plusOne([4, 3, 2, 1])) ; // [4, 3, 2, 2]
echo "</pre>";

echo "<pre>plusOne([9, 9, 9])\r\n";
var_dump(plusOne([9, 9, 9])) ; // [1, 0, 0, 0]
echo "</pre>";

File: Plus One/plus-one.go
package main

import (
	"fmt"
)

func plusOne(digits []int) []int {
	for i := len(digits) - 1; i >= 0; i-- {
		if digits[i] < 9 {
			digits[i]++
			return digits
		}

		digits[i] = 0
		fmt.Print(i)
	}

	if digits[0] == 0 {
		digits = append([]int{1}, digits...)
	}

	return digits
}

func main() {
	arr1 := []int{1, 2, 3}
	arr2 := []int{4, 3, 2, 1}
	arr3 := []int{9, 9, 9}

	fmt.Printf("plusOne([1, 2, 3]) = %v\n", plusOne(arr1)) // [1, 2, 4]
	fmt.Printf("plusOne([4, 3, 2, 1]) = %v\n", plusOne(arr2)) // [4, 3, 2, 1]
	fmt.Printf("plusOne([9, 9, 9]) = %v\n", plusOne(arr3)) // [1, 0, 0, 0]
}